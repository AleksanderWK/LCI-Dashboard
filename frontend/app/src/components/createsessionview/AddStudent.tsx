import React, {useEffect, useState} from "react";
import {createStyles, makeStyles, Theme, Typography, TextField, Button, CardActions} from "@material-ui/core";
import AddIcon from "@material-ui/icons/Add";
import {useRecoilState, useSetRecoilState} from "recoil";
import {
    addStudentPopupOpenState,
    createSessionValuesState
} from "../../state/CreateSessionViewState/createSessionViewAtoms";
import {students, User} from "../../state/data/studentAtoms";
import {ipcGet, ipcSend} from "../../ipc";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        form: {
            display: "grid",
            gridTemplateColumns: "1fr",
            gap: 36,
            color: theme.palette.text.default
        },
        inputNewStudentName: {
            width: "100%"
        },
        btn: {
            color: theme.palette.text.default,
            fontWeight: "bold"
        }
    })
);

export default function AddStudent(): JSX.Element {
    const classes = useStyles();

    const [studentName, setStudentName] = useState<string>("");
    const [inputError, setInputError] = useState<boolean>(false);
    const setPopupOpen = useSetRecoilState(addStudentPopupOpenState);
    const setCreateSessionValues = useSetRecoilState(createSessionValuesState);
    const [student_list, setStudentList] = useRecoilState(students);

    const addStudent = () => {
        // If student is in list already prevent action and show error on screen
        if (student_list.includes(studentName) || studentName == "") {
            setInputError(true);
        } else {
            // Add student, choose the student and close popup
            ipcSend("insertUser", {name: studentName});
            const newList = student_list.concat(studentName);
            setStudentList(newList);
            setCreateSessionValues((prevValues) => ({
                ...prevValues,
                studentName: studentName
            }));
            setPopupOpen(false);
        }
    };

    // Remove error message if user starts rewriting error-triggering student name.
    useEffect(() => {
        setInputError(false);
        ipcGet<User[]>("getUsers").then((users) => {
            setStudentList(users.map((user) => user.name));
        });
    }, [studentName]);

    return (
        <div className={classes.form}>
            <Typography variant="h1" style={{color: "#000000"}}>
                Add Student
            </Typography>
            <TextField
                className={classes.inputNewStudentName}
                label={"Student name"}
                inputProps={{"data-testid": "content-input"}}
                onChange={(event: React.ChangeEvent<{value: unknown}>) => {
                    setStudentName(event.target.value as string);
                }}
                error={inputError}
            />
            <CardActions style={{width: "100%", justifyContent: "flex-end"}}>
                <Button
                    onClick={() => {
                        setPopupOpen(false);
                    }}
                >
                    Cancel
                </Button>

                <Button
                    variant="contained"
                    color="primary"
                    onClick={addStudent}
                    className={classes.btn}
                    startIcon={<AddIcon />}
                >
                    Add student
                </Button>
            </CardActions>
        </div>
    );
}
