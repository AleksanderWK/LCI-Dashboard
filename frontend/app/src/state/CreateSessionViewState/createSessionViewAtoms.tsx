import {atom} from "recoil";

export interface CreateSessionValues {
    sessionName: string;
    studentName: string;
    eyeTracker: string;
}

export const createSessionValuesState = atom<CreateSessionValues>({
    key: "createSessionValues",
    default: {
        sessionName: "",
        studentName: "",
        eyeTracker: ""
    }
});

export const addStudentPopupOpenState = atom<boolean>({
    key: "addStudentPopupOpen",
    default: false
});
