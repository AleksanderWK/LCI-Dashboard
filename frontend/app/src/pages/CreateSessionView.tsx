import React, {Suspense} from "react";
import {makeStyles, createStyles} from "@material-ui/core";
import logo from "../assets/Images/LCI_logo.png";
import dashboardIllustration from "../assets/Images/dashboardIllustration.svg";
import CreateSession from "../components/createsessionview/CreateSession";
import PopupContainer from "../components/common/PopupContainer";
import AddStudent from "../components/createsessionview/AddStudent";
import {useRecoilState} from "recoil";
import {addStudentPopupOpenState} from "../state/popup";
import Popup from "../components/common/Popup";

const useStyles = makeStyles(() =>
    createStyles({
        pageContainer: {
            width: "100%",
            height: "100%",
            display: "grid",
            gridTemplateColumns: "30px 140px auto 140px 30px",
            gridTemplateRows: "30px auto 30px"
        },
        pageLogo: {
            gridColumnStart: 2,
            gridColumnEnd: 3,
            gridRowStart: 2,
            gridRowEnd: 3
        },
        pageContent: {
            display: "flex",
            justifyContent: "center",
            alignItems: "center",
            width: "100%",
            height: "100%",
            gridColumnStart: 3,
            gridColumnEnd: 4,
            gridRowStart: 2,
            gridRowEnd: 3
        },
        illustration: {
            width: "400px",
            height: "400px",
            marginLeft: "100px"
        }
    })
);

export default function CreateSessionView(): JSX.Element {
    const classes = useStyles();

    const [addStudentPopupOpen, setAddStudentPopupOpen] = useRecoilState(addStudentPopupOpenState);

    return (
        <>
            <div className={classes.pageContainer}>
                <img className={classes.pageLogo} src={logo} alt="The LCI-lab logo" />
                <div className={classes.pageContent}>
                    <Suspense fallback={<div>Loading...</div>}>
                        <CreateSession />
                    </Suspense>
                    <img className={classes.illustration} src={dashboardIllustration} alt="dashboard illustration" />
                </div>
            </div>

            <PopupContainer
                open={addStudentPopupOpen}
                onClose={() => {
                    setAddStudentPopupOpen(false);
                }}
            >
                <Popup>
                    <AddStudent />
                </Popup>
            </PopupContainer>
        </>
    );
}
