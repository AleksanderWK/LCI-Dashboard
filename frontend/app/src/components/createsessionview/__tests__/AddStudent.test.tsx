import React from "react";
import ReactDOM from "react-dom";
import {fireEvent, render, screen, waitFor, getByTestId} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import AddStudent from "../AddStudent";
import {RecoilRoot} from "recoil";
import {addStudentPopupOpenState} from "../../../state/CreateSessionViewState/createSessionViewAtoms";
import CreateSessionView from "../../../pages/CreateSessionView";

it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
        <RecoilRoot
            initializeState={(snap) => {
                snap.set(addStudentPopupOpenState, true);
            }}
        >
            <AddStudent />
        </RecoilRoot>,
        div
    );
});

it("AddStudentPopup matches snapshot", () => {
    const {baseElement} = render(
        <RecoilRoot
            initializeState={(snap) => {
                snap.set(addStudentPopupOpenState, true);
            }}
        >
            <AddStudent />
        </RecoilRoot>
    );
    expect(baseElement).toMatchSnapshot();
});

it("cant add empty string", async () => {
    render(
        <RecoilRoot
            initializeState={(snap) => {
                snap.set(addStudentPopupOpenState, true);
            }}
        >
            <AddStudent />
        </RecoilRoot>
    );
    const nameField = screen.getByRole("textbox", {hidden: true});
    const submitBtn = screen.getByRole("button", {
        name: /add student/i,
        hidden: true
    });
    await fireEvent.click(submitBtn);
    expect(nameField.parentElement).toHaveClass("Mui-error");
});

it("cant add same student twice", async () => {
    render(
        <RecoilRoot
            initializeState={(snap) => {
                snap.set(addStudentPopupOpenState, true);
            }}
        >
            <CreateSessionView />
        </RecoilRoot>
    );
    const nameField = screen.getByTestId("content-input");
    const submitBtn = screen.getByRole("button", {
        name: /add student/i,
        hidden: true
    });
    const popupBtn = screen.getByRole("button", {name: /add new student/i});
    await fireEvent.change(nameField, {target: {value: "Aleksander"}});
    await fireEvent.click(submitBtn);
    await fireEvent.click(popupBtn);
    await fireEvent.click(submitBtn);
    expect(nameField.parentElement).toHaveClass("Mui-error");
});
/*
it("popup adds student to selection dropdown", async () => {
    render(
        <RecoilRoot
            initializeState={(snap) => {
                snap.set(popupOpen, true);
            }}
        >
            <CreateSessionView />
        </RecoilRoot>
    );
    const nameField = screen.getByTestId("content-input");
    const submitBtn = screen.getByRole("button", {
        name: /add student/i,
        hidden: true
    });
    const dropdownField = screen.getByLabelText("Student");
    await fireEvent.change(nameField, {target: {value: "Aleksander"}});
    await fireEvent.click(submitBtn);
    await fireEvent.click(dropdownField);
    expect(screen.getByTestId("Aleksander")).not.toThrowError();
});
//test: popup adds student to selection dropdown (expect option to be in document)
*/
