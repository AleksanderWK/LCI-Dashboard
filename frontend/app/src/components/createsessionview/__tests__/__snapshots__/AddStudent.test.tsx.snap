// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddStudentPopup matches snapshot 1`] = `
<body>
  <div>
    <div
      class="makeStyles-form-1"
    >
      <h1
        class="MuiTypography-root MuiTypography-h1"
        style="color: rgb(0, 0, 0);"
      >
        Add Student
      </h1>
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-inputNewStudentName-2"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
        >
          Student name
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            data-testid="content-input"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
        style="width: 100%; justify-content: flex-end;"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-btn-3"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Cancel
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-btn-3 MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeMedium"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
            </span>
            Add student
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
